src/server/pfs/server/driver.go

```diff
@@ -1917,6 +1919,60 @@ func (d *driver) putFile(ctx context.Context, file *pfs.File, delimiter pfs.Deli
 				value = jsonValue
 			case pfs.Delimiter_LINE:
 				value, err = bufioR.ReadBytes('\n')
+			case pfs.Delimiter_SQL:
+				rows := int64(1)
+				if targetFileDatums != 0 {
+					rows = targetFileDatums
+				}
+				_value, n, _err := sqlReader.ReadRows(rows)
+				if _err == nil || _err == io.EOF {
+					if _err == io.EOF {
+						// Now that we're done reading, populate the header/footer
+						// records
+						headerBuffer := &bytes.Buffer{}
+						headerBuffer.Write(sqlReader.Header)
+						limiter.Acquire()
+						eg.Go(func() error {
+							defer limiter.Release()
+							object, size, err := pachClient.PutObject(headerBuffer)
+							if err != nil {
+								return err
+							}
+							mu.Lock()
+							defer mu.Unlock()
+							records.Header = &pfs.PutFileRecord{
+								SizeBytes:  size,
+								ObjectHash: object.Hash,
+							}
+							return nil
+						})
+						footerBuffer := &bytes.Buffer{}
+						footerBuffer.Write(sqlReader.Footer)
+						limiter.Acquire()
+						eg.Go(func() error {
+							defer limiter.Release()
+							object, size, err := pachClient.PutObject(footerBuffer)
+							if err != nil {
+								return err
+							}
+							mu.Lock()
+							defer mu.Unlock()
+							records.Footer = &pfs.PutFileRecord{
+								SizeBytes:  size,
+								ObjectHash: object.Hash,
+							}
+							return nil
+						})
+					}
+
+					if n > 0 {
+						// only increment here if the count is greater than 0,
+						// since we increment by default a few lines below
+						datumsWritten += n - int64(1)
```

Comment:
The only reason you have to do this is because your code doesn't fit in with the code around it. Each of the other cases reads a single instance of the delimited value off of the wire. Yours is reading multiple which is where the problems are coming from. Make it read a single value, and everything will make a lot more sense. Alternatively you could make the outside code expect to read multiple datums in each iteration of the loop... but I think that would be worse.

Reply:
Ah, somehow I missed that. Yes, this is much much simpler now.